File: multi_robot_py_topics_pkg/multi_robot_py_topics_pkg/multi_robot_subscriber.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String
from multi_robot_interfaces_pkg.msg import AddressBook
from sensor_msgs.msg import Image # Image is the message type
from cv_bridge import CvBridge # Package to convert between ROS and OpenCV Images
import cv2 # OpenCV library

class MultiRobotPyImageSubscriberClass(Node):
    """
    Create an ImageSubscriber class, which is a subclass of the Node class.
    """
    def __init__(self):
        """
        Class constructor to set up the node
        """
        # Initiate the Node class's constructor and give it a name
        super().__init__('multi_robot_image_subscriber_node')
          
        # Create the subscriber. This subscriber will receive an Image
        # from the video_frames topic. The queue size is 10 messages.
        self.subscription = self.create_subscription(
          Image, 
          'video_frames', 
          self.listener_callback, 
          10)
        self.subscription # prevent unused variable warning
          
        # Used to convert between ROS and OpenCV images
        self.br = CvBridge()
   
    def listener_callback(self, data):
        """
        Callback function.
        """
        # Display the message on the console
        self.get_logger().info('Receiving video frame')
     
        # Convert ROS Image message to OpenCV image
        current_frame = self.br.imgmsg_to_cv2(data,"bgr8")
        
        # Display image
        cv2.imshow("camera", current_frame)
        
        cv2.waitKey(1)

class MultiRobotPyStringSubscriberClass(Node):

    def __init__(self):
        super().__init__('multi_robot_string_subscriber_node')
        self.std_queue_size_ = 10
        self.subscription_std_ = self.create_subscription(
            String,
            'multi_robot_std_topic',
            self.std_listener_callback,
            self.std_queue_size_)
        self.subscription_std_  # prevent unused variable warning

        self.custom_queue_size_ = 10
        self.subscription_custom_ = self.create_subscription(
            AddressBook,
            'multi_robot_custom_topic',
            self.custom_listener_callback,
            self.custom_queue_size_)
        self.subscription_custom_  # prevent unused variable warning

    def std_listener_callback(self, msg):
        self.get_logger().info('I heard: "%s"' % msg.data)


    def custom_listener_callback(self, msg):
        self.get_logger().info('I heard: "%s"' % msg.first_name)

def main(args=None):
    rclpy.init(args=args)

    multi_robot_py_image_subscriber_obj = MultiRobotPyImageSubscriberClass()

    rclpy.spin(multi_robot_py_image_subscriber_obj)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    multi_robot_py_image_subscriber_obj.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: multi_robot_cpp_topics_pkg/src/multi_robot_subscriber.cpp
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "multi_robot_interfaces_pkg/msg/address_book.hpp"

using std::placeholders::_1;

class MultiRobotCppSubscriber : public rclcpp::Node
{
  public:
    MultiRobotCppSubscriber()
    : Node("multi_robot_subscriber_node", rclcpp::NodeOptions().allow_undeclared_parameters(true).automatically_declare_parameters_from_overrides(true)), std_queue_size_(10), custom_queue_size_(10)
    {

    std::string std_topic_name = this->get_parameter("std_topic_name").as_string();

    std::string custom_topic_name = this->get_parameter("custom_topic_name").as_string();


      // Create two subscribers, one each for multi_robot_std_topic and multi_robot_custom_topic topics
      std_subscription_ = this->create_subscription<std_msgs::msg::String>(
      std_topic_name, std_queue_size_, std::bind(&MultiRobotCppSubscriber::std_topic_callback, this, _1));

      custom_subscription_ = this->create_subscription<multi_robot_interfaces_pkg::msg::AddressBook>(
      custom_topic_name, custom_queue_size_, std::bind(&MultiRobotCppSubscriber::custom_topic_callback, this, _1));
    }

  private:
    // callback function for the multi_robot_std_topic subscriber
    void std_topic_callback(const std_msgs::msg::String & msg) const
    {
      RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg.data.c_str());
    }

    // callback function for the multi_robot_custom_topic subscriber
    void custom_topic_callback(const multi_robot_interfaces_pkg::msg::AddressBook & msg) const
    {
      RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg.first_name.c_str());
    }

    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr std_subscription_;
    rclcpp::Subscription<multi_robot_interfaces_pkg::msg::AddressBook>::SharedPtr custom_subscription_;
    size_t std_queue_size_, custom_queue_size_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MultiRobotCppSubscriber>());
  rclcpp::shutdown();
  return 0;
}